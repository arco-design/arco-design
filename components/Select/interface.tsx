import { ReactNode, CSSProperties, PropsWithChildren, ReactElement, HTMLAttributes } from 'react';
import { TriggerProps } from '../Trigger';
import { SelectViewCommonProps } from '../_class/select-view';
import { AvailableVirtualListProps } from '../_class/VirtualList';

export type OptionsType = SelectProps['options'];

// 造成输入框值改变的原因：用户输入、选中选项、选项下拉框收起、触发自动分词
export type InputValueChangeReason =
  | 'manual'
  | 'optionChecked'
  | 'optionListHide'
  | 'tokenSeparator';

export interface OptionInfo extends PropsWithChildren<OptionProps> {
  child?: ReactElement;
  _valid: boolean;
  _index: number;
  _origin: 'children' | 'options' | 'userCreatedOptions' | 'userCreatingOption';
}

export type LabeledValue = {
  value: string | number;
  label: ReactNode;
};

export type SelectInnerStateValue = string | number | string[] | number[];

/**
 * @title Select
 */
export interface SelectProps extends SelectViewCommonProps {
  /**
   * @zh 选择框的默认值
   * @en To set default value
   */
  defaultValue?: string | string[] | number | number[] | LabeledValue | LabeledValue[];
  /**
   * @zh 选择器的值（受控模式）
   * @en To set value
   */
  value?: string | string[] | number | number[] | LabeledValue | LabeledValue[];
  /**
   * @zh 输入框的值（受控模式）
   * @en To set input value
   */
  inputValue?: string;
  /**
   * @zh 是否开启多选模式或标签模式 (**`tags` 推荐使用 `mode: multiple; allowCreate: true` 替代，下一大版本将移除此模式**)
   * @en Set mode of Select(**`tags` recommends using `mode: multiple; allowCreate: true` instead, this mode will be removed in the next major version**)
   */
  mode?: 'multiple' | 'tags';
  /**
   * @zh 指定可选项
   * @en Select options
   * @version `extra` in 2.2.0
   */
  options?: (
    | string
    | number
    | { label: ReactNode | string; value: string | number; disabled?: boolean; extra?: any }
  )[];
  /**
   * @zh 设置 `onChange` 回调中 `value` 的格式。默认是string，设置为true时候，value格式为： { label: string, value: string }
   * @en Whether to embed label in value, turn the format of value from string to `{ value: string, label: ReactNode }`
   */
  labelInValue?: boolean;
  /**
   * @zh 是否根据输入的值筛选数据。如果传入函数的话，接收 `inputValue` 和 `option` 两个参数，当option符合筛选条件时，返回 `true`，反之返回 `false`。
   * @en If it's true, filter options by input value. If it's a function, filter options base on the function.
   * @defaultValue true
   */
  filterOption?: boolean | ((inputValue: string, option: ReactElement) => boolean);
  /**
   * @zh
   * 定制回显内容。返回值将会显示在下拉框内。若 `value` 对应的 `Option` 不存在，则第一个参数是 null
   * @en
   * Customize the content that will be displayed in the Select.
   * If the `Option` corresponding to `value` does not exist, the first parameter will be `null`
   */
  renderFormat?: (option: OptionInfo | null, value: string | number | LabeledValue) => ReactNode;
  /**
   * @zh 是否默认高亮第一个选项
   * @en Whether to highlight the first option by default
   * @defaultValue true
   */
  defaultActiveFirstOption?: boolean;
  /**
   * @zh 是否在隐藏的时候销毁 DOM 结构
   * @en Whether to destroy the DOM when hiding
   * @defaultValue true
   */
  unmountOnExit?: boolean;
  /**
   * @zh 下拉框是否默认打开。
   * @en Whether to show dropdown by default.
   * @version 2.14.0
   */
  defaultPopupVisible?: boolean;
  /**
   * @zh 控制下拉框是否打开。
   * @en Whether to show dropdown.
   * @version 2.6.0
   */
  popupVisible?: boolean;
  /**
   * @zh 没有数据时显示的内容
   * @en Specify content to show when no result matches.
   */
  notFoundContent?: ReactNode;
  /**
   * @zh 在多选模式下自动分词的分隔符。
   * @en Separator used to tokenize on `multiple` mode.
   */
  tokenSeparators?: string[];
  /**
   * @zh 弹出框挂载的父节点。
   * @en To set the container of the dropdown.
   */
  getPopupContainer?: (node: HTMLElement) => Element;
  /**
   * @zh 触发方式。
   * @en The trigger mode which executes the dropdown action.
   * @defaultValue click
   */
  trigger?: TriggerProps['trigger'];
  /**
   * @zh 自定义触发元素。
   * @en The trigger element which executes the dropdown action.
   * @version `() => ReactNode` in 2.31.0
   */
  triggerElement?:
    | ReactNode
    | ((params: { value: any; option: OptionInfo | OptionInfo[] }) => ReactNode);
  /**
   * @zh 可以接受所有 `Trigger` 的 `Props`
   * @en Pass all `Trigger` component properties
   */
  triggerProps?: Partial<TriggerProps>;
  /**
   * @zh 自定义弹出内容。
   * @en Customize dropdown content
   */
  dropdownRender?: (menu: ReactNode) => ReactNode;
  /**
   * @zh 下拉列表的样式。
   * @en The style of dropdown menu.
   */
  dropdownMenuStyle?: CSSProperties;
  /**
   * @zh 下拉列表的类。
   * @en The className of dropdown menu.
   */
  dropdownMenuClassName?: string | string[];
  /**
   * @zh 传递虚拟滚动属性。
   * @en Pass properties used by VirtualList.
   * @version 2.1.0
   */
  virtualListProps?: AvailableVirtualListProps;
  /**
   * @zh 点击选择框的回调
   * @en Callback when select an option or input value change.
   */
  onChange?: (value, option: OptionInfo | OptionInfo[]) => void;
  /**
   * @zh 取消选中的时候触发的回调，(只在 `multiple` 模式下触发)。
   * @en Called when a option is deselected.Only called for `multiple` mode.
   */
  onDeselect?: (value: string | number | LabeledValue, option: OptionInfo) => void;
  /**
   * @zh 点击清除时触发，参数是当前下拉框的展开状态。
   * @en Called when clear
   */
  onClear?: (visible: boolean) => void;
  /**
   * @zh 搜索时的回调
   * @en Callback when input changed
   */
  onSearch?: (value: string, reason: InputValueChangeReason) => void;
  /**
   * @zh 获得焦点时的回调
   * @en Callback when get focus
   */
  onFocus?: (e) => void;
  /**
   * @zh 失去焦点时的回调
   * @en Callback when lose focus
   */
  onBlur?: (e) => void;
  /**
   * @zh 下拉框的滚动监听函数，参数为滚动元素。
   * @en Callback when dropdown scrolls.
   */
  onPopupScroll?: (elem) => void;
  /**
   * @zh 下拉框收起展开时触发
   * @en Callback when visibility of dropdown is changed.
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * @zh 输入框文本改变的回调。
   * @en Callback when the value of input is changed.
   * @version 2.3.0
   */
  onInputValueChange?: (value: string, reason: InputValueChangeReason) => void;
  /**
   * @zh 输入框文本粘贴的回调。
   * @en Callback when the you paste text in input box.
   * @version 2.9.0
   */
  onPaste?: (e) => void;
}

/**
 * @title Select.Option
 */
export interface OptionProps extends Omit<HTMLAttributes<HTMLLIElement>, 'className'> {
  _key?: any;
  style?: CSSProperties;
  children?: ReactNode;
  prefixCls?: string;
  rtl?: boolean;
  className?: string | string[];
  wrapperClassName?: string | string[];
  /**
   * @zh 是否禁用
   * @en Whether the option is disabled
   */
  disabled?: boolean;
  /**
   * @zh 默认根据此属性值进行筛选
   * @en Value of this Option
   */
  value: string | number;
  /**
   * @zh 携带任意自定义数据。
   * @en Any data you want to pass to Option.
   * @version 2.2.0
   */
  extra?: any;
  // Some user may use isSelectOption to hack, Do NOT change it to _isSelectOption
  isSelectOption?: boolean;
  _isMultipleMode?: boolean;
  _isUserCreatedOption?: boolean;
  _isUserCreatingOption?: boolean;
  _valueActive?: OptionProps['value'];
  _valueSelect?: SelectInnerStateValue;
  _onClick?: (value: OptionProps['value'], disabled: boolean) => void;
  _onMouseEnter?: (value: OptionProps['value']) => void;
  _onMouseLeave?: () => void;
}

/**
 * @title Select.OptGroup
 */
export interface OptGroupProps extends HTMLAttributes<HTMLLIElement> {
  _key?: any;
  children?: ReactNode;
  prefixCls?: string;
  /**
   * @zh 组名
   * @en Name of Group
   */
  label?: ReactNode;
  // Some user may use isSelectOptGroup to hack, Do NOT change it to _isSelectOptGroup
  isSelectOptGroup?: boolean;
}

/**
 * @title Select Reference Type
 */
export type SelectHandle = {
  /**
   * @zh DOM 节点
   * @en DOM
   */
  dom: HTMLDivElement;
  /**
   * @zh 使选择框聚焦
   * @en Focus Select
   */
  focus: () => void;
  /**
   * @zh 使选择框失焦
   * @en Blur Select
   */
  blur: () => void;
  /**
   * @zh 鼠标快捷操作的处理函数
   * @en Processor of mouse shortcut operation
   */
  hotkeyHandler: (event: KeyboardEvent) => void;
  /**
   * @zh 处于悬浮态的选项的值
   * @en The value of active option
   */
  activeOptionValue: OptionProps['value'];
  /**
   * @zh 获得选项信息的列表
   * @en Get the list of option info
   */
  getOptionInfoList: () => OptionInfo[];
  /**
   * @zh 根据选项值获得对应的选项信息
   * @en Get the option info by its value
   */
  getOptionInfoByValue: (value: OptionProps['value']) => OptionInfo;
  /**
   * @zh 将下拉列表滚动至指定选项
   * @en Scroll the drop-down list to the specified option
   * @version 2.46.0
   */
  scrollIntoView: (value: OptionProps['value'], options?: ScrollIntoViewOptions) => void;
};
