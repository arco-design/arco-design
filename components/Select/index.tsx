import Select from './select';

export default Select;

export type { SelectProps, OptionProps, OptGroupProps } from './interface';
