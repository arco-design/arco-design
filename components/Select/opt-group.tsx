import React from 'react';
import { OptGroupProps } from './interface';
import omit from '../_util/omit';

function OptGroup(props: OptGroupProps, ref) {
  const { prefixCls, label, ...rest } = props;
  return (
    <li
      ref={ref}
      className={`${prefixCls}-group-title`}
      {...omit(rest, ['_key', 'children', 'isSelectOptGroup'])}
    >
      {label}
    </li>
  );
}

const ForwardRefOptGroup = React.forwardRef<unknown, OptGroupProps>(OptGroup);

const OptGroupComponent = ForwardRefOptGroup as typeof ForwardRefOptGroup & {
  __ARCO_SELECT_OPTGROUP__?: boolean;
};

OptGroupComponent.__ARCO_SELECT_OPTGROUP__ = true;

export default OptGroupComponent;
