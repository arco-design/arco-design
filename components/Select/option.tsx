import React from 'react';
import cs from '../_util/classNames';
import Checkbox from '../Checkbox';
import { OptionProps } from './interface';
import omit from '../_util/omit';

function Option(props: OptionProps, ref) {
  const {
    style,
    className,
    wrapperClassName,
    disabled,
    prefixCls,
    rtl,
    value: propValue,
    children: propChildren,
    _isMultipleMode,
    _isUserCreatedOption,
    _isUserCreatingOption,
    _valueActive,
    _valueSelect,
    _onMouseEnter,
    _onMouseLeave,
    _onClick,
    ...rest
  } = props;

  const value = 'value' in props ? propValue : `${propChildren}`;
  const childNode = 'children' in props ? propChildren : `${propValue}`;

  const isChecked = _isMultipleMode
    ? (_valueSelect as any[]).indexOf(value) !== -1
    : _valueSelect === value;

  const optionLabelProps = {
    style,
    className: cs(
      `${prefixCls}-option`,
      {
        [`${prefixCls}-option-selected`]: isChecked,
        [`${prefixCls}-option-disabled`]: disabled,
        [`${prefixCls}-option-hover`]: value === _valueActive,
        [`${prefixCls}-option-empty`]:
          (!childNode && childNode !== 0) ||
          (typeof childNode === 'string' && /^\s*$/.test(childNode)),
        [`${prefixCls}-option-rtl`]: rtl,
      },
      className
    ),
    onMouseEnter: (event) => {
      _onMouseEnter && _onMouseEnter(value);
      rest.onMouseEnter && rest.onMouseEnter(event);
    },
    onMouseLeave: (event) => {
      _onMouseLeave && _onMouseLeave();
      rest.onMouseLeave && rest.onMouseLeave(event);
    },
    onClick: (event) => {
      _onClick && _onClick(value, disabled);
      rest.onClick && rest.onClick(event);
    },
    ...omit(rest, ['_key', 'extra', 'isSelectOption', 'onClick', 'onMouseEnter', 'onMouseLeave']),
  };

  const wrapperProps = {
    ref,
    role: 'option',
    'aria-selected': isChecked,
  };
  // Mark the option that created/creating by user self
  _isUserCreatedOption && Object.assign(wrapperProps, { 'data-user-created': true });
  _isUserCreatingOption && Object.assign(wrapperProps, { 'data-user-creating': true });

  if (_isMultipleMode) {
    return (
      <li
        {...wrapperProps}
        className={cs(
          `${prefixCls}-option-wrapper`,
          {
            [`${prefixCls}-option-wrapper-selected`]: isChecked,
            [`${prefixCls}-option-wrapper-disabled`]: disabled,
          },
          wrapperClassName
        )}
      >
        <Checkbox
          aria-hidden="true"
          className={`${prefixCls}-checkbox`}
          checked={isChecked}
          disabled={disabled}
          onChange={optionLabelProps.onClick}
        />
        <span {...optionLabelProps}>{childNode}</span>
      </li>
    );
  }

  return (
    <li {...wrapperProps} {...optionLabelProps}>
      {childNode}
    </li>
  );
}

const ForwordRefOption = React.forwardRef<unknown, OptionProps>(Option);

const OptionComponent = ForwordRefOption as typeof ForwordRefOption & {
  __ARCO_SELECT_OPTION__?: boolean;
};

OptionComponent.__ARCO_SELECT_OPTION__ = true;

export default OptionComponent;
