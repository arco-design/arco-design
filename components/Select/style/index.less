@import './token.less';
@import './mixin.less';

@select-prefix-cls: ~'@{prefix}-select';
@select-prefix-cls-rtl: ~'@{prefix}-select-rtl';

.select-view(@select-prefix-cls);

.@{select-prefix-cls} {
  &-popup {
    top: 4px;
    box-sizing: border-box;
    padding: @select-popup-padding-vertical @select-popup-padding-horizontal;
    border: 1px solid @select-popup-color-border;
    border-radius: @select-popup-border-radius;
    background-color: @select-popup-color-bg;
    box-shadow: @select-popup-box-shadow;
    overflow: hidden;

    &-hidden {
      display: none;
    }
  }

  &-popup &-popup-inner {
    width: 100%;
    max-height: @select-popup-max-height;
    list-style: none;
  }

  &-popup &-option {
    position: relative;
    box-sizing: border-box;
    width: 100%;
    padding: 0 @select-signal-popup-option-padding-horizontal;
    font-size: @select-popup-font-size;
    text-align: left;
    cursor: pointer;
    line-height: @select-popup-option-height;
    border-radius: @select-popup-option-border-radius;

    .text-ellipsis();

    .option-color(@status) {
      @class-suffix: if(@status = default, ~'', ~'-@{status}');

      &@{class-suffix} {
        color: ~'@{select-popup-option-color-text_@{status}}';
        background-color: ~'@{select-popup-option-color-bg_@{status}}';
      }
    }

    .option-color(default);
    .option-color(selected);
    .option-color(hover);
    // disabled 优先级最高，放在最后
    .option-color(disabled);

    &-disabled {
      cursor: not-allowed;
    }

    &-selected {
      font-weight: @select-popup-option-font-weight_selected;
    }

    &-empty {
      height: $line-height;
    }
  }

  &-popup &-option-rtl {
    text-align: right;
  }

  &-option-wrapper {
    display: flex;
    align-items: center;
    padding: 0 7px;

    .@{select-prefix-cls}-checkbox {
      padding: 0 5px;

      // 避免点击复选框导致 select-view 的输入框失焦
      input {
        display: none;
      }
    }

    .@{select-prefix-cls}-option {
      flex: 1;
      margin-left: 1px;
      padding: 0 @select-multi-popup-option-padding-horizontal;
      border-radius: @select-border-radius;
    }
  }

  &-group-title {
    box-sizing: border-box;
    width: 100%;
    padding: @select-popup-group-title-padding-top @select-popup-group-title-padding-horizontal 0
      @select-popup-group-title-padding-horizontal;
    line-height: @select-popup-group-title-height;
    font-size: @select-popup-group-title-font-size;
    color: @select-popup-group-title-color-text;
    .text-ellipsis();

    &:first-child {
      padding-top: @select-popup-group-title-padding-top - @select-popup-padding-vertical;
    }
  }

  &-highlight {
    font-weight: @select-popup-option-font-hightlight-weight;
    color: @select-popup-option-color-hightlight-text;
  }
}

.@{select-prefix-cls}-rtl {
  direction: rtl;
}

.select-view-rtl(@select-prefix-cls, @select-prefix-cls-rtl);
