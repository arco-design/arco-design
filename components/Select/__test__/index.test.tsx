import React from 'react';
import { mount } from 'enzyme';
import { sleep, $ } from '../../../tests/util';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Button from '../../Button';
import Select from '../select';
import { Enter, Tab } from '../../_util/keycode';
import { LabeledValue } from '../interface';

mountTest(Select);
componentConfigTest(Select, 'Select');

function mountSelect(component) {
  return mount(component);
}

const Option = Select.Option;
const options = [
  '南非龙虾',
  '新西兰羊排',
  '海鲜烩意面',
  '摩卡慕斯',
  '酱烧豆腐',
  '千舟一叶',
  '葱油鲷鱼',
  '提拉米苏',
  '奶油蘑菇鸡肉卷',
];

let wrapper = null;

describe('Select', () => {
  beforeEach(() => {
    wrapper && wrapper.unmount();
  });

  it('tokenSeparators', async () => {
    const onChange = jest.fn();
    wrapper = mountSelect(
      <Select
        popupVisible
        allowCreate
        mode="multiple"
        tokenSeparators={[',', ';', '\n']}
        onChange={onChange}
      />
    );

    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value: 'a,b' } })
      .simulate('keydown', {
        key: Enter.key,
      });

    expect(onChange.mock.calls[0][0]).toEqual(['a', 'b']);

    await sleep(100);

    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value: 'c' } })
      .simulate('keydown', {
        key: ',',
      })
      .simulate('change', { target: { value: 'c,' } });

    expect(wrapper.find('.arco-select-option')).toHaveLength(3);
  });

  it('tokenSeparators with allowCreate is false', async () => {
    wrapper = mountSelect(
      <Select
        mode="multiple"
        placeholder="Please select"
        tokenSeparators={[',', '|', '/', '\n']}
        allowCreate={false}
        allowClear
        style={{ width: 345 }}
      />
    );

    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value: 'a,b' } })
      .simulate('keydown', {
        key: Enter.key,
      });

    expect(wrapper.find('.arco-tag').exists()).toBeFalsy();

    await sleep(100);

    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value: 'c' } })
      .simulate('keydown', {
        key: Tab.key,
      });

    expect(wrapper.find('.arco-tag').exists()).toBeFalsy();
  });

  it('OptGroup works', () => {
    wrapper = mountSelect(
      <Select popupVisible options={[1, 2, 3]}>
        <Select.OptGroup label="GroupOne">
          <Select.Option value="G1_V1">G1_V1</Select.Option>
          <Select.Option value="G1_V2">G1_V2</Select.Option>
        </Select.OptGroup>
      </Select>
    );

    expect(wrapper.find('.arco-select-group-title').at(0).text()).toBe('GroupOne');
    expect(wrapper.find('.arco-select-option')).toHaveLength(5);
  });

  it('onChange listener correctly', () => {
    const mockFn = jest.fn();
    jest.useFakeTimers();
    wrapper = mountSelect(
      <Select
        placeholder="请选择语言"
        style={{ width: 300 }}
        defaultValue="南非龙虾"
        allowClear
        onChange={mockFn}
      >
        {options.map((option) => (
          <Option key={option} value={option}>
            {option}
          </Option>
        ))}
      </Select>
    );
    expect(wrapper.find('input').props().value).toBe('南非龙虾');

    wrapper.simulate('click');
    jest.runAllTimers();
    expect($('.arco-select-popup').length).toBe(1);

    wrapper.find('.arco-select-option').at(2).simulate('click');
    expect(wrapper.find('input').props().value).toBe('海鲜烩意面');
    expect(mockFn.mock.calls.length).toBe(1);
  });

  it('multiply mode correctly,clear correctly, disabled correctly', () => {
    const onChange = jest.fn();
    wrapper = mountSelect(
      <Select
        mode="multiple"
        placeholder="请选择语言"
        style={{ width: 300 }}
        defaultValue={['新西兰羊排', '酱烧豆腐', '海鲜烩意面']}
        allowClear
        onChange={onChange}
      >
        {options.map((option, index) => (
          <Option key={option} value={option} disabled={index === 2}>
            {option}
          </Option>
        ))}
      </Select>
    );
    expect(wrapper.find('.arco-tag')).toHaveLength(3);

    wrapper.find('.arco-select').at(0).simulate('click');
    expect($('.arco-select-popup-inner .arco-select-option-disabled').length).toBe(1);

    wrapper.find('IconClose').last().simulate('click');
    expect(onChange.mock.calls[0][0]).toHaveLength(1);
  });

  it('showSearch correctly', () => {
    wrapper = mountSelect(
      <Select showSearch allowClear placeholder="请选择食物" style={{ width: 300 }}>
        {options.map((option) => (
          <Option key={option} value={option}>
            {option}
          </Option>
        ))}
      </Select>
    );
    const input = wrapper.find('input');

    wrapper.find('.arco-select').at(0).simulate('click');
    expect(
      wrapper
        .find('.arco-select-popup .arco-select-option')
        .at(0)
        .hasClass('arco-select-option-hover')
    ).toBeTruthy();

    input.simulate('change', { target: { value: '海鲜' } });

    expect(wrapper.find('.arco-select-popup .arco-select-option')).toHaveLength(1);
    expect(wrapper.find('.arco-select-popup .arco-select-option').at(0).text()).toEqual(
      '海鲜烩意面'
    );

    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value: '@' } });
    expect(wrapper.find('Empty')).toHaveLength(1);
  });

  it('showSearch async correctly', () => {
    const mockFn = jest.fn();
    wrapper = mountSelect(
      <Select
        mode="multiple"
        placeholder="请选择语言"
        style={{ width: 300 }}
        onSearch={mockFn}
        filterOption={false}
        defaultValue={['新西兰羊排', '酱烧豆腐', '海鲜烩意面']}
        allowClear
      >
        {options.map((option, index) => (
          <Option key={option} value={option} disabled={index === 2}>
            {option}
          </Option>
        ))}
      </Select>
    );
    wrapper.find('.arco-select').at(0).simulate('click');
    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value: '南' } });
    expect(mockFn.mock.calls[0][0]).toBe('南');
  });

  it('popOver correctly', () => {
    wrapper = mountSelect(
      <Select
        placeholder="请选择语言"
        style={{ width: 300 }}
        defaultValue="南非龙虾"
        allowClear
        triggerProps={{
          position: 'left',
        }}
      >
        {options.map((option) => (
          <Option key={option} value={option}>
            {option}
          </Option>
        ))}
      </Select>
    );
    wrapper.find('.arco-select').at(0).simulate('click');
    expect(wrapper.find('.arco-trigger').hasClass('arco-trigger-position-left')).toBeTruthy();
  });

  it('show value correctly when defaultValue does not exist in options', () => {
    wrapper = mountSelect(
      <Select className="test" mode="multiple" defaultValue={[1]}>
        {options.map((option, index) => (
          <Option key={option} value={option} disabled={index === 2}>
            {option}
          </Option>
        ))}
      </Select>
    );
    expect(wrapper.find('.arco-tag')).toHaveLength(1);
    expect(wrapper.find('.arco-tag').at(0).text()).toBe('1');

    const singleWrapper = mountSelect(
      <Select defaultValue={1}>
        {options.map((option, index) => (
          <Option key={option} value={option} disabled={index === 2}>
            {option}
          </Option>
        ))}
      </Select>
    );
    expect(singleWrapper.find('input').prop('value')).toBe(1);
  });

  it('renderFormat correctly', () => {
    wrapper = mountSelect(
      <Select
        showSearch
        allowClear
        defaultValue={1}
        placeholder="请选择食物"
        renderFormat={(option) => {
          return `${option.value} - ${option.children}`;
        }}
        options={options.map((option, index) => ({ label: option, value: index }))}
      />
    );
    const input = wrapper.find('input');
    expect(input.prop('value')).toBe(`1 - ${options[1]}`);
    wrapper.find('.arco-select-view').simulate('click');
    // expect(wrapper.find('input').prop('value')).toBeFalsy();
    expect(input.prop('placeholder')).toBe(`1 - ${options[1]}`);
  });

  it('dropdownRender', () => {
    wrapper = mountSelect(
      <Select
        dropdownRender={(menu) => {
          return (
            <div>
              {menu}
              <Button>Add</Button>
            </div>
          );
        }}
        options={options}
      />
    );
    wrapper.find('.arco-select').at(0).simulate('click');
    expect(wrapper.find('.arco-select-popup .arco-btn').text()).toBe('Add');
  });

  it('dropdownStyle & dropdownClassName', () => {
    wrapper = mountSelect(
      <Select options={options} dropdownMenuStyle={{ fontSize: 20 }} dropdownMenuClassName="xxx" />
    );
    wrapper.find('.arco-select').at(0).simulate('click');

    const menu = wrapper.find('.arco-select-popup .arco-select-popup-inner').at(0);
    expect(menu.hasClass('xxx')).toBeTruthy();
    expect(getComputedStyle(menu.getDOMNode()).getPropertyValue('font-size')).toBe('20px');
  });

  it('Compatible when labelInValue is set, value is passed in the object', () => {
    wrapper = mountSelect(
      <Select
        mode="multiple"
        labelInValue
        options={[
          { label: '1', value: 1 },
          { label: '2', value: 2 },
        ]}
      />
    );

    expect(wrapper.find('.arco-tag')).toHaveLength(0);

    wrapper = mountSelect(
      <Select
        mode="multiple"
        labelInValue
        defaultValue={[{ label: '1', value: 1 }, 2] as any}
        options={[
          { label: '1', value: 1 },
          { label: '2', value: 2 },
        ]}
      />
    );

    expect(wrapper.find('.arco-tag')).toHaveLength(2);

    wrapper = mountSelect(
      <Select
        placeholder="Please select"
        labelInValue
        value={{ value: 1, label: 'One' }}
        renderFormat={(_, labeledValue) => {
          const { label, value } = labeledValue as LabeledValue;
          return label || value;
        }}
      />
    );

    expect(wrapper.find('.arco-select-view-value').at(0).text()).toBe('One');

    wrapper = mountSelect(
      <Select
        placeholder="Please select"
        labelInValue
        mode="multiple"
        defaultValue={[{ value: 1, label: 'One' }]}
        renderFormat={(_, labeledValue) => {
          const { label, value } = labeledValue as LabeledValue;
          return label || value;
        }}
      />
    );

    expect(wrapper.find('.arco-input-tag-tag-content').at(0).text()).toBe('One');
  });

  it('User creating option will not affect original options', () => {
    const value = 'Hello';
    wrapper = mountSelect(
      <Select
        allowCreate
        popupVisible
        mode="multiple"
        inputValue={value}
        defaultValue={[value]}
        options={[{ value, label: value, disabled: true }]}
      />
    );
    expect(wrapper.find('.arco-select-option-disabled')).toHaveLength(1);
  });
});
