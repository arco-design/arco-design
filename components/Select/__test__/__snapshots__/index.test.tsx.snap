// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Select default 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-select arco-select-single arco-select-size-default"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-select-view"
    title=""
  >
    <span
      class="arco-select-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-select-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-select-view-value arco-select-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-select-suffix"
    >
      <div
        class="arco-select-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Select set className globally 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-select arco-select-single arco-select-size-default global-default-class"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-select-view"
    title=""
  >
    <span
      class="arco-select-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-select-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-select-view-value arco-select-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-select-suffix"
    >
      <div
        class="arco-select-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Select set className globally and component set className itself. 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-select arco-select-single arco-select-size-default component-class"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-select-view"
    title=""
  >
    <span
      class="arco-select-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-select-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-select-view-value arco-select-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-select-suffix"
    >
      <div
        class="arco-select-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Select set data-* property 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-select arco-select-single arco-select-size-default"
  data-test-name="Select"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-select-view"
    title=""
  >
    <span
      class="arco-select-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-select-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-select-view-value arco-select-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-select-suffix"
    >
      <div
        class="arco-select-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
